datafusion/optimizer/src/extract_equijoin_predicate.rs

```diff
@@ -0,0 +1,438 @@
+    use crate::test::*;
+
+    use super::*;
+
+    fn assert_optimized_plan_eq(plan: &LogicalPlan, expected: Vec<&str>) {
+        let rule = ExtractEquijoinPredicate::new();
+        let optimized_plan = rule
+            .try_optimize(plan, &OptimizerContext::new())
+            .unwrap()
+            .expect("failed to optimize plan");
+        let formatted = optimized_plan.display_indent_schema().to_string();
+        let actual: Vec<&str> = formatted.trim().lines().collect();
+
+        assert_eq!(
+            expected, actual,
+            "\n\nexpected:\n\n{:#?}\nactual:\n\n{:#?}\n\n",
+            expected, actual
+        );
+
+        assert_eq!(plan.schema(), optimized_plan.schema())
+    }
+
+    #[test]
+    fn join_with_only_column_equi_predicate() -> Result<()> {
+        let t1 = test_table_scan_with_name("t1")?;
+        let t2 = test_table_scan_with_name("t2")?;
+
+        let plan = LogicalPlanBuilder::from(t1)
+            .join(
+                t2,
+                JoinType::Left,
+                (Vec::<Column>::new(), Vec::<Column>::new()),
+                Some(col("t1.a").eq(col("t2.a"))),
+            )?
+            .build()?;
+        let expected = vec![
+            "Left Join: t1.a = t2.a [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
+            "  TableScan: t1 [a:UInt32, b:UInt32, c:UInt32]",
+            "  TableScan: t2 [a:UInt32, b:UInt32, c:UInt32]",
+        ];
+        assert_optimized_plan_eq(&plan, expected);
+
+        Ok(())
+    }
+
+    #[test]
+    fn join_with_only_equi_expr_predicate() -> Result<()> {
+        let t1 = test_table_scan_with_name("t1")?;
+        let t2 = test_table_scan_with_name("t2")?;
+
+        let plan = LogicalPlanBuilder::from(t1)
+            .join(
+                t2,
+                JoinType::Left,
+                (Vec::<Column>::new(), Vec::<Column>::new()),
+                Some((col("t1.a") + lit(10i64)).eq(col("t2.a") * lit(2u32))),
+            )?
+            .build()?;
+        let expected = vec![
+            "Left Join: t1.a + Int64(10) = t2.a * UInt32(2) [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
+            "  TableScan: t1 [a:UInt32, b:UInt32, c:UInt32]",
+            "  TableScan: t2 [a:UInt32, b:UInt32, c:UInt32]",
+        ];
+        assert_optimized_plan_eq(&plan, expected);
+
+        Ok(())
+    }
+
+    #[test]
+    fn join_with_only_none_equi_predicate() -> Result<()> {
+        let t1 = test_table_scan_with_name("t1")?;
+        let t2 = test_table_scan_with_name("t2")?;
+
+        let plan = LogicalPlanBuilder::from(t1)
+            .join(
+                t2,
+                JoinType::Left,
+                (Vec::<Column>::new(), Vec::<Column>::new()),
+                Some(binary_expr(
+                    (col("t1.a") + lit(10i64)).gt_eq(col("t2.a") * lit(2u32)),
+                    Operator::And,
+                    col("t1.b").lt(lit(100i32)),
+                )),
```

Comment:
I thunk you can use `.and` here if you want (untested)

```suggestion
                Some(
                    col("t1.a") + lit(10i64)).gt_eq(col("t2.a") * lit(2u32))
                      .and(col("t1.b").lt(lit(100i32)))
                )),
```

Reply:
It is short and clear, fixed.